{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'primer1': {'forward primer': 'TATAGCATTCCCCCGATTGA',\n",
      "             'reverse primer': 'CACTCCCGCAAGATGTAATG'},\n",
      " 'primer2': {'forward primer': 'CATTACATCTTGCGGGAGTG',\n",
      "             'reverse primer': 'ATAGCACCAGCCAGAACAGG'},\n",
      " 'primer3': {'forward primer': 'TATAGCATTCCCCCGATTGA',\n",
      "             'reverse primer': 'TCCCGCAAGATGTAATGAAA'},\n",
      " 'primer4': {'forward primer': 'CATTACATCTTGCGGGAGTG',\n",
      "             'reverse primer': 'GAATTGGGTCTCCACCTCCT'},\n",
      " 'primer5': {'forward primer': 'TATAGCATTCCCCCGATTGA',\n",
      "             'reverse primer': 'ACACTCCCGCAAGATGTAATG'}}\n"
     ]
    }
   ],
   "source": [
    "#ライブラリ読み込み\n",
    "from IPython.display import clear_output\n",
    "from selenium import webdriver\n",
    "import time\n",
    "from pprint import pprint\n",
    "from Bio import AlignIO\n",
    "#相対パスを指定\n",
    "\n",
    "file_path = 'Sympetrum_consensus/Sympetrum_kunckeli_consensus.fasta'\n",
    "target = AlignIO.read(file_path, 'fasta')\n",
    "print(target.format('fasta'))\n",
    "\n",
    "\n",
    "#クロームの立ち上げ\n",
    "driver=webdriver.Chrome()\n",
    "\n",
    "#URLの指定\n",
    "driver.get(\"https://www.bioinformatics.nl/cgi-bin/primer3plus/primer3plus.cgi\")\n",
    "time.sleep(1)\n",
    "\n",
    "#検索ボックスの指定\n",
    "search_box=driver.find_element_by_xpath(\"/html/body/div/div[2]/form/div[5]/div[3]/table[1]/tbody/tr[2]/td/textarea\")\n",
    "\n",
    "#検索ボックスに文字を入力\n",
    "search_box.send_keys(target.format('fasta'))\n",
    "\n",
    "#検索ボタンの指定\n",
    "search_button=driver.find_element_by_xpath(\"/html/body/div/div[2]/form/div[2]/table/tbody/tr/td[3]/table/tbody/tr/td[1]/input\")\n",
    "\n",
    "#検索ボタンをクリック\n",
    "search_button.click()\n",
    "time.sleep(0.5)\n",
    "\n",
    "#ブラウザのウインドウ高を取得する\n",
    "win_height = driver.execute_script(\"return window.innerHeight\")\n",
    "\n",
    "#スクロール開始位置の初期値（ページの先頭からスクロールを開始する）\n",
    "last_top = 1\n",
    "\n",
    "#ページの最下部までスクロールする無限ループ\n",
    "while True:\n",
    "\n",
    "  #スクロール前のページの高さを取得\n",
    "  last_height = driver.execute_script(\"return document.body.scrollHeight\")\n",
    "  \n",
    "  #スクロール開始位置を設定\n",
    "  top = last_top\n",
    "\n",
    "  #ページ最下部まで、徐々にスクロールしていく\n",
    "  while top < last_height:\n",
    "    top += int(win_height * 0.1)\n",
    "    driver.execute_script(\"window.scrollTo(0, %d)\" % top)\n",
    "    time.sleep(0.05)\n",
    "\n",
    "  #１秒待って、スクロール後のページの高さを取得する\n",
    "  time.sleep(1)\n",
    "  new_last_height = driver.execute_script(\"return document.body.scrollHeight\")\n",
    "\n",
    "  #スクロール前後でページの高さに変化がなくなったら無限スクロール終了とみなしてループを抜ける\n",
    "  if last_height == new_last_height:\n",
    "    break\n",
    "\n",
    "  #次のループのスクロール開始位置を設定\n",
    "  last_top = last_height\n",
    "\n",
    "# primer dict\n",
    "primers = {}\n",
    "i = 0\n",
    "x = 0\n",
    "while True:\n",
    "    try:\n",
    "        forward_primer = driver.find_element_by_id(f\"PRIMER_{i}_SEQUENCE\").get_attribute(\"value\")\n",
    "        i += 1\n",
    "        x += 1\n",
    "        reverse_primer = driver.find_element_by_id(f\"PRIMER_{i}_SEQUENCE\").get_attribute(\"value\")\n",
    "        primers[f\"primer{x}\"] = {\"forward primer\": forward_primer, \"reverse primer\": reverse_primer}\n",
    "        i += 1\n",
    "    except:\n",
    "        break\n",
    "\n",
    "clear_output(wait=True)\n",
    "pprint(primers)\n",
    "\n",
    "#クロームの終了\n",
    "driver.quit()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Needleman-Wunschアルゴリズムについて\n",
    "https://bi.biopapyrus.jp/seq/alignment/needleman%E2%80%93wunsch.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mprimer1's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer1's forward primer is found in related species sequences.\n",
      "\u001b[31mprimer1's forward primer is not found in related species sequences.\u001b[0m\n",
      "\u001b[31mprimer1's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer1's reverse primer is found in related species sequences.\n",
      "primer1's reverse primer is found in related species sequences.\n",
      "primer1's reverse primer is found in related species sequences.\n",
      "\u001b[31mprimer1's reverse primer is not found in related species sequences.\u001b[0m\n",
      "primer2's forward primer is found in related species sequences.\n",
      "primer2's forward primer is found in related species sequences.\n",
      "primer2's forward primer is found in related species sequences.\n",
      "\u001b[31mprimer2's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer2's reverse primer is found in related species sequences.\n",
      "primer2's reverse primer is found in related species sequences.\n",
      "primer2's reverse primer is found in related species sequences.\n",
      "\u001b[31mprimer2's reverse primer is not found in related species sequences.\u001b[0m\n",
      "\u001b[31mprimer3's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer3's forward primer is found in related species sequences.\n",
      "\u001b[31mprimer3's forward primer is not found in related species sequences.\u001b[0m\n",
      "\u001b[31mprimer3's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer3's reverse primer is found in related species sequences.\n",
      "primer3's reverse primer is found in related species sequences.\n",
      "primer3's reverse primer is found in related species sequences.\n",
      "\u001b[31mprimer3's reverse primer is not found in related species sequences.\u001b[0m\n",
      "primer4's forward primer is found in related species sequences.\n",
      "primer4's forward primer is found in related species sequences.\n",
      "primer4's forward primer is found in related species sequences.\n",
      "\u001b[31mprimer4's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer4's reverse primer is found in related species sequences.\n",
      "\u001b[31mprimer4's reverse primer is not found in related species sequences.\u001b[0m\n",
      "primer4's reverse primer is found in related species sequences.\n",
      "\u001b[31mprimer4's reverse primer is not found in related species sequences.\u001b[0m\n",
      "\u001b[31mprimer5's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer5's forward primer is found in related species sequences.\n",
      "\u001b[31mprimer5's forward primer is not found in related species sequences.\u001b[0m\n",
      "\u001b[31mprimer5's forward primer is not found in related species sequences.\u001b[0m\n",
      "primer5's reverse primer is found in related species sequences.\n",
      "primer5's reverse primer is found in related species sequences.\n",
      "primer5's reverse primer is found in related species sequences.\n",
      "\u001b[31mprimer5's reverse primer is not found in related species sequences.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq\n",
    "#候補に上がったprimerと、target種の配列データを比較する\n",
    "#file_path = '種特異にしたい種の配列データ'のパス\n",
    "file_path = '/home/dkcom/デスクトップ/primer_design/alignment_python/Sympetrum/Sympetrum_kunckeli.fasta'\n",
    "target = AlignIO.read(file_path, 'fasta')\n",
    "\n",
    "# 配列データの比較\n",
    "for primer_num, primer in primers.items():\n",
    "    for record in target:\n",
    "        for i in range(len(record.seq) - len(primer[\"forward primer\"])):\n",
    "            alignments = pairwise2.align.globalxx(primer[\"forward primer\"], str(record.seq[i:i+len(primer[\"forward primer\"])]))\n",
    "            if alignments[0][2] / alignments[0][4] >= 1.0:\n",
    "                print(f\"{primer_num}'s forward primer is found in related species sequences.\")\n",
    "                break\n",
    "        else:\n",
    "            print(\"\\033[31m{}\\033[0m\".format(f\"{primer_num}'s forward primer is not found in related species sequences.\"))\n",
    "    # reverse primerを相補鎖に変換\n",
    "    rev_primer = Seq(primer[\"reverse primer\"]).reverse_complement()\n",
    "    for record in target: \n",
    "        for i in range(len(record.seq) - len(rev_primer)):\n",
    "            alignments = pairwise2.align.globalxx(rev_primer, str(record.seq[i:i+len(rev_primer)]))\n",
    "            if alignments[0][2] / alignments[0][4] >= 1.0:\n",
    "                print(f\"{primer_num}'s reverse primer is found in related species sequences.\")\n",
    "                break\n",
    "        else:\n",
    "            print(\"\\033[31m{}\\033[0m\".format(f\"{primer_num}'s reverse primer is not found in related species sequences.\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "primer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15 (default, Nov 24 2022, 15:19:38) \n[GCC 11.2.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2bb0823a418acd0601b4af9cd4086f1e18910aea704b27735c1d1f1cab4a9f35"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
